gapic/src/main/com/google/gapid/models/Devices.java

```diff
@@ -193,9 +194,37 @@ public void selectReplayDevice(Device.Instance dev) {
     listeners.fire().onReplayDeviceChanged(dev);
   }
 
-  public ListenableFuture<DeviceValidationResult> validateDevice(Device.Instance device) {
-    return transformAsync(client.validateDevice(Paths.device(device.getID())), r ->
-      submitIfNotDisposed(shell, () -> validationCache.add(device, new DeviceValidationResult(r))));
+  public void validateDevice(Device.Instance device, Consumer<DeviceValidationResult> onValidationResultCb) {
+    rpcController.start().listen(client.validateDevice(Paths.device(device.getID())),
+        new UiErrorCallback<Service.DeviceValidationResult, Service.DeviceValidationResult, RpcException>(shell, LOG) {
+      @Override
+      protected ResultOrError<Service.DeviceValidationResult, RpcException>
+        onRpcThread(Rpc.Result<Service.DeviceValidationResult> response) throws RpcException, ExecutionException {
+        try {
+          return success(response.get());
+        } catch (RpcException e) {
+          // Expected to get RPC exceptions for internal errors.
+          return error(e);
+        } catch (ExecutionException e) {
+          throttleLogRpcError(LOG, "LoadData error", e);
+          return error(null);
+        }
+      }
+
+      @Override
+      protected void onUiThreadSuccess(Service.DeviceValidationResult r) {
+        DeviceValidationResult result = new DeviceValidationResult(r);
+        submitIfNotDisposed(shell, () -> validationCache.add(device, result));
```

Comment:
No need to schedule this on the UI thread, since you already are on the UI thread here.